### PR TITLE
add RandomSampler

### PR DESCRIPTION
**Pull request by gracew**
This change is [Reviewable](https://reviewable.io/reviews/palantir/http-remoting/423)

### CODE REVIEW

File: tracing/src/main/java/com/palantir/remoting2/tracing/RandomSampler.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.Random;
+
+public final class RandomSampler implements TraceSampler {
+
+    private final float rate;
+    private final Random random;
+
+    public RandomSampler(float rate) {
```
Comment (markelliot) [inline tracing/src/main/java/com/palantir/remoting2/tracing/RandomSampler.java]: should add docs

File: tracing/src/main/java/com/palantir/remoting2/tracing/RandomSampler.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        checkArgument(rate > 0 && rate < 1, "rate should be between 0 and 1: was %s", rate);
+        this.rate = rate;
+        this.random = new Random();
```
Comment (markelliot) [inline tracing/src/main/java/com/palantir/remoting2/tracing/RandomSampler.java]: should probably use a ThreadLocalRandom

Comment (markelliot) [reply tracing/src/main/java/com/palantir/remoting2/tracing/RandomSampler.java]: ...and just make this a static

**Review by uschi2000 — APPROVED** [review]

File: tracing/src/main/java/com/palantir/remoting2/tracing/RandomSampler.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    private final float rate;
+
+    public RandomSampler(float rate) {
+        checkArgument(rate >= 0 && rate <= 1, "rate should be between 0 and 1: was %s", rate);
+        this.rate = rate;
+    }
+
+    @Override
+    public boolean sample() {
+        return random.nextFloat() < rate;
```
Comment (schlosna) [inline tracing/src/main/java/com/palantir/remoting2/tracing/RandomSampler.java:40]: I don't think we want this in a static field, as it effectively serializes all threads through a single random that was set when this class is loaded, see https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadLocalRandom.html

Instead we need to call current each time to get the appropriate thread local random for the current thread: 

`return ThreadLocalRandom.current().nextFloat() < rate;`

Comment (gracew) [reply tracing/src/main/java/com/palantir/remoting2/tracing/RandomSampler.java:40]: https://github.com/palantir/http-remoting/pull/426

Comment (schlosna) [reply tracing/src/main/java/com/palantir/remoting2/tracing/RandomSampler.java:40]: Thanks @gracew

**Pull request merged**